source/common/config/config_provider_impl.cc

```diff
@@ -39,31 +61,34 @@ bool ConfigSubscriptionInstanceBase::checkAndApplyConfig(const Protobuf::Message
     // bindConfigProvider()).
     // This makes it safe to call any of the provider's onConfigProtoUpdate() to get a new config
     // impl, which can then be passed to all providers.
+    auto* typed_provider = static_cast<MutableConfigProviderBase*>(provider);
     if (new_config == nullptr) {
-      if ((new_config = provider->onConfigProtoUpdate(config_proto)) == nullptr) {
+      if ((new_config = typed_provider->onConfigProtoUpdate(config_proto)) == nullptr) {
         return false;
       }
     }
-    provider->onConfigUpdate(new_config);
+    typed_provider->onConfigUpdate(new_config);
   }
 
   return true;
 }
 
-void ConfigSubscriptionInstanceBase::bindConfigProvider(MutableConfigProviderImplBase* provider) {
-  // All config providers bound to a ConfigSubscriptionInstanceBase must be of the same concrete
-  // type; this is assumed by checkAndApplyConfig() and is verified by the assertion below.
-  // NOTE: an inlined statement ASSERT() triggers a potentially evaluated expression warning from
-  // clang due to `typeid(**mutable_config_providers_.begin())`. To avoid this, we use a lambda to
-  // separate the first mutable provider dereference from the typeid() statement.
-  ASSERT([&]() {
-    if (!mutable_config_providers_.empty()) {
-      const auto& first_provider = **mutable_config_providers_.begin();
-      return typeid(*provider) == typeid(first_provider);
-    }
-    return true;
-  }());
-  mutable_config_providers_.insert(provider);
+void DeltaConfigSubscriptionInstance::applyConfigUpdate(
+    const std::function<void(const ConfigSharedPtr&)>& updateFn) {
+  // The Config implementation is assumed to be shared across the config providers bound to this
+  // subscription, therefore, simply propagating the update to all worker threads for a single bound
+  // provider will be sufficient.
+  if (mutable_config_providers_.size() > 1) {
+    ASSERT(static_cast<DeltaMutableConfigProviderBase*>(*mutable_config_providers_.begin())
+               ->getConfig() == static_cast<DeltaMutableConfigProviderBase*>(
+                                    *std::next(mutable_config_providers_.begin()))
+                                    ->getConfig());
+  }
+
+  auto* typed_provider =
+      static_cast<DeltaMutableConfigProviderBase*>(getAnyBoundMutableConfigProvider());
+  ConfigSharedPtr config = typed_provider->getConfig();
+  typed_provider->onConfigUpdate([config, updateFn]() { updateFn(config); });
```

Comment:
So, in the delta case, we don't check to see if the new resources are already in the vector? How do we avoid reapplying the same config twice?

Reply:
I added a TODO here. This logic can definitely be generalized, although I need to inspect a few more use cases (CDS, SRDS, possibly LDS?) including the changes to support delta subscriptions to better understand the path to a common implementation.

The parent PR has the TODO for this in the scoped RDS [onConfigUpdate()](https://github.com/envoyproxy/envoy/pull/5839/files#diff-1578b366b91c6312bf7d120645d32b22R131).

Preferably, I would like to get this PR in with the TODO so I can unblock the follow up PRs at which point I can focus on framework improvements while @stevenzzzz focuses on the scoped routing logic.